### ISSUE TITLE
Issues with OpenJDK/OpenJFX, Maven, sound in linux, white box, malfunction/medical events, tutorial

### ISSUE THREAD

**Issue by shirishag75** [mars-sim/mars-sim#1]
mars-sim throws an exception with openjdk 8 

```
> mars-sim                                                                                                                         [95%]
11 Jan, 2017 2:26:36 PM(INFO) main.javafx.MarsProjectFX Mars Simulation Project v3.1.0 - Build 3837 - Java 8.0.111 (Thu Nov 24 14:50:47 UTC 2016) - 4 CPU thread(s)
Exception in Application init method
java.lang.reflect.InvocationTargetException
	at sun.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
	at sun.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:62)
	at sun.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
	at java.lang.reflect.Method.invoke(Method.java:498)
	at com.sun.javafx.application.LauncherImpl.launchApplicationWithArgs(LauncherImpl.java:389)
	at com.sun.javafx.application.LauncherImpl.launchApplication(LauncherImpl.java:328)
	at sun.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
	at sun.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:62)
	at sun.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
	at java.lang.reflect.Method.invoke(Method.java:498)
	at sun.launcher.LauncherHelper$FXHelper.main(LauncherHelper.java:767)
Caused by: java.lang.RuntimeException: Exception in Application init method
	at com.sun.javafx.application.LauncherImpl.launchApplication1(LauncherImpl.java:912)
	at com.sun.javafx.application.LauncherImpl.lambda$launchApplication$1(LauncherImpl.java:182)
	at java.lang.Thread.run(Thread.java:745)
Caused by: java.lang.ArrayIndexOutOfBoundsException: 3
	at org.mars_sim.msp.javafx.MarsProjectFX.init(MarsProjectFX.java:280)
	at com.sun.javafx.application.LauncherImpl.launchApplication1(LauncherImpl.java:841)
	... 2 more
Exception running application org.mars_sim.msp.javafx.MarsProjectFX
```

Hope the above helps.

**Comment by mokun** [comment]
I can experiment with setting up openJDK 8 and openJFK 8 in Eclipse under my windows os to see if they can compile the existing codebase. 

If not, we'll have to wait for openJDK 9. 

I was Oracle's JDK 9 will introduce some "breaking" changes in the internal API such as how JavaFX's CSS styles are processed and so forth. We may want to delay the adoption of JDK 9 or openJDK 9 even after it is released in order to ensure the code conversion is properly done.

**Comment by shirishag75** [comment]
were you able to get something out of it ?

**Comment by mokun** [comment]
I'm using mostly Windows OS as platform for my work.

I go to [ojdkbuild](https://github.com/ojdkbuild/ojdkbuild) for a ready-to-go windows binary build of OpenJDK 8

The trouble is with finding a windows binary of OpenJFX 8. 

I haven't got to the point in which I can compile OpenJFX myself on my machine.

I just found this [site](https://github.com/SkyLandTW/OpenJFX-binary-windows) for OpenJFX 8u60.

Let see if they will work together to replace Oracle Java 8.

Questions: 

Do you already have both OpenJFX 8 and OpenJDK 8 installed on your linux box ?

With them, can you run mars-sim ? and do you have the stack traces to show ?

**Comment by shirishag75** [comment]
@mokun I can answer some of the questions, some I'll need inputs from you to seek and do -

As far as OpenJFX8 and OpenJDK 8 are concerned, I have both . In fact I'm in the midst of upgrading both -

```
[$] apt-cache policy openjdk-8-jdk openjdk-8-jre
                                                                                    
openjdk-8-jdk:
  Installed: 8u111-b14-3
  Candidate: 8u121-b13-2
  Version table:
     8u121-b13-2 600
        600 http://httpredir.debian.org/debian stretch/main amd64 Packages
          1 http://httpredir.debian.org/debian unstable/main amd64 Packages
 *** 8u111-b14-3 100
        100 /var/lib/dpkg/status
openjdk-8-jre:
  Installed: 8u111-b14-3
  Candidate: 8u121-b13-2
  Version table:
     8u121-b13-2 600
        600 http://httpredir.debian.org/debian stretch/main amd64 Packages
          1 http://httpredir.debian.org/debian unstable/main amd64 Packages
 *** 8u111-b14-3 100
        100 /var/lib/dpkg/status
```

Although openjfx is also not far behind -

```
[$] apt-cache policy openjfx libopenjfx-java libopenjfx-jni 
                                                                        
openjfx:
  Installed: 8u111-b14-1
  Candidate: 8u111-b14-1
  Version table:
     8u121-b13-2 1
          1 http://httpredir.debian.org/debian unstable/main amd64 Packages
 *** 8u111-b14-1 600
        600 http://httpredir.debian.org/debian stretch/main amd64 Packages
        100 /var/lib/dpkg/status
libopenjfx-java:
  Installed: 8u111-b14-1
  Candidate: 8u111-b14-1
  Version table:
     8u121-b13-2 1
          1 http://httpredir.debian.org/debian unstable/main amd64 Packages
     8u121-b13-1 1
          1 http://httpredir.debian.org/debian unstable/main i386 Packages
 *** 8u111-b14-1 600
        600 http://httpredir.debian.org/debian stretch/main amd64 Packages
        600 http://httpredir.debian.org/debian stretch/main i386 Packages
        100 /var/lib/dpkg/status
libopenjfx-jni:
  Installed: 8u111-b14-1
  Candidate: 8u111-b14-1
  Version table:
     8u121-b13-2 1
          1 http://httpredir.debian.org/debian unstable/main amd64 Packages
 *** 8u111-b14-1 600
        600 http://httpredir.debian.org/debian stretch/main amd64 Packages
        100 /var/lib/dpkg/status
```

What's your setup and can you run mars-sim ? 

By my setup if you mean hardware - the current system I'm working on is pretty old system. It's a wolfdale dual core processor (45nm socket) .  2 GB DDR2 RAM,

**Comment by shirishag75** [comment]
[hs_err_pid21230.log.zip](https://github.com/mars-sim/mars-sim/files/740765/hs_err_pid21230.log.zip)

Output on CLI -

```
[$] mars-sim                                                                                                                        [4:52:57]
Jan 31, 2017 4:53:02 AM(INFO) main.javafx.MarsProjectFX Mars Simulation Project v3.1.0 - Build 3817 - Java SE 8.0.111 (Thu Nov 24 14:50:47 UTC 2016) - 2 CPU thread(s)
Jan 31, 2017 4:53:02 AM(INFO) main.javafx.MarsProjectFX Starting a new sim in GUI mode in linux
Jan 31, 2017 4:53:07 AM(INFO) ui.javafx.mainmenu.SpinningGlobe JavaFX 3D features supported
Jan 31, 2017 5:02:05 AM(INFO) ui.javafx.mainmenu.MainMenu Creating a new sim in linux
Setting Alpha Base's Objective to Crop Farm
Setting Schiaparelli Point's Objective to Crop Farm
Jan 31, 2017 5:05:22 AM(SEVERE) core.time.UpTimer Too long since last time pulse, clearing update time
Jan 31, 2017 5:05:23 AM(INFO) ui.javafx.MainScene MainScene's initializeScene() is on JavaFX Application Thread Thread
Jan 31, 2017 5:05:44 AM(INFO) core.person.ai.task.ProduceFood ChefBot 006 has an accident while performing food production.
Jan 31, 2017 5:05:44 AM(INFO) core.malfunction.MalfunctionManager An accident occurs in a Lounge
Jan 31, 2017 5:05:44 AM(INFO) core.malfunction.Malfunction Electrical Short - the repair requires the part(s) electrical wire (quantity: 1)
Jan 31, 2017 5:05:44 AM(INFO) ui.javafx.MainScene MainScene's createJFXTabs() is on JavaFX Application Thread Thread
Jan 31, 2017 5:05:44 AM(INFO) ui.javafx.MainScene MainScene's createFXZoomSlider() is on JavaFX Application Thread Thread
Jan 31, 2017 5:05:44 AM(INFO) ui.javafx.MainScene MainScene's prepareOthers() is on JavaFX Application Thread Thread
Jan 31, 2017 5:05:47 AM(SEVERE) ui.swing.sound.OGGSoundClip Unable to open the sourceDataLine: javax.sound.sampled.LineUnavailableException
java.lang.IllegalArgumentException: Master Gain not supported
	at org.classpath.icedtea.pulseaudio.PulseAudioLine.getControl(PulseAudioLine.java:89)
	at org.classpath.icedtea.pulseaudio.PulseAudioSourceDataLine.getControl(PulseAudioSourceDataLine.java:51)
	at org.mars_sim.msp.ui.swing.sound.OGGSoundClip.setGain(OGGSoundClip.java:168)
	at org.mars_sim.msp.ui.swing.sound.OGGSoundClip.initJavaSound(OGGSoundClip.java:466)
	at org.mars_sim.msp.ui.swing.sound.OGGSoundClip.getOutputLine(OGGSoundClip.java:482)
	at org.mars_sim.msp.ui.swing.sound.OGGSoundClip.playStream(OGGSoundClip.java:619)
	at org.mars_sim.msp.ui.swing.sound.OGGSoundClip.access$000(OGGSoundClip.java:34)
	at org.mars_sim.msp.ui.swing.sound.OGGSoundClip$2.run(OGGSoundClip.java:391)
#
# A fatal error has been detected by the Java Runtime Environment:
#
#  SIGSEGV (0xb) at pc=0x00007ff12d112ef8, pid=24519, tid=0x00007ff14a04f700
#
# JRE version: OpenJDK Runtime Environment (8.0_111-b14) (build 1.8.0_111-8u111-b14-3-b14)
# Java VM: OpenJDK 64-Bit Server VM (25.111-b14 mixed mode linux-amd64 compressed oops)
# Problematic frame:
# C  [libjfxwebkit.so+0x1883ef8]
#
# Core dump written. Default location: /home/shirish/games/core or core.24519
#
# An error report file with more information is saved as:
# /home/shirish/games/hs_err_pid24519.log
#
# If you would like to submit a bug report, please visit:
#   http://bugreport.java.com/bugreport/crash.jsp
# The crash happened outside the Java Virtual Machine in native code.
# See problematic frame for where to report the bug.
#
Aborted (core dumped)
```

The only difference between the errors are the log numbers, other than that, the errors are same.

**Comment by shirishag75** [comment]
![mars-configuration-editor](https://cloud.githubusercontent.com/assets/727626/22446906/418f161a-e775-11e6-9b0f-23eabf906ad8.png)

Default configuration editor, I didn't change a thing - just clicked on the bright arrow in the center.

**Comment by mokun** [comment]
> ui.swing.sound.OGGSoundClip Unable to open the sourceDataLine: javax.sound.sampled.LineUnavailableException

This tells me your sound card is not being set up properly. Can you double check ?

I thought I once turned off looking for sound card in the past. But it seems that mars-sim is still too sensitive to any error in sound setting in linux. 

Meanwhile, I'll compile a new jar for you to test on your machine and will send you a link soon.

btw, my openJDK and openJFX binary versions are not matched up. This seems to be the reason that  throws me errors as I run mars-sim on top of it in windows.

If you can find the compiled windows binary of OpenJFX based on 8u91+, by all means, do you let me know the link.

**Comment by mokun** [comment]
This build is based on r3848 and compressed with 7z : 

for debian,
see https://www.dropbox.com/s/z0cp44hzxbs24pi/r3848deb.7z?dl=0

for jar,
see https://www.dropbox.com/s/agwujjbnrkqpcen/r3848.7z?dl=0

**Comment by mokun** [comment]
To build OpenJFX, see https://wiki.openjdk.java.net/display/OpenJFX/Building+OpenJFX

**Comment by larsnaesbye** [comment]
Did this solve the problem?

**Comment by mokun** [comment]
@larsnaesbye, do you sound issue or openjdk/openjfx support ?

I don't have a fast linux box setup. Ecipse is not as responsive in linux than in windows. The graphic support is notorious in linux. This is one reason I stick with windows. 

So far I don't have the right openjdk/openjfx version for testing in windows.

I guess we should talk about just openjdk/openjfx on this thread and separate the sound configuration into a new issue. 

> > ui.swing.sound.OGGSoundClip Unable to open the sourceDataLine: javax.sound.sampled.LineUnavailableException

> This tells me your sound card is not being set up properly. Can you double check ?

**Comment by larsnaesbye** [comment]
I can get sound fine both in Windows and Linux, or at least I could when I last tested that some months ago.

**Comment by mokun** [comment]
I've just resolved the sound issue in linux in the latest build [3894](https://github.com/mars-sim/mars-sim/commit/dd37004ca9cfc8518ba4135a65e291df9c14d3a8) 

This morning I managed to set up ubuntu 17.04 on my laptop. 

@shirishag75 I installed OpenJDK 8 with OpenJFX and did not use Oracle's JDK 8.

@larsnaesbye The fix is nothing fancy. I simply disabled mars-sim's the sound slider bar due to the
incompatibility problem of pulseaudio in the jorbis library with whatever ubuntu is using for sound config. As a result, user may use linux's built-in sound panel to control the volume of mars-sim's background music and sound effect. 

If you take a look, you will see I made some changes in OGGSoundClip's setGain(), added disableSound() in MainScene and added and made use of a new boolean field hasMasterGain in AudioPlayer, etc.

**Comment by mokun** [comment]
@shirishag75 I've found that mars-sim is capable of running under ubuntu with OpenJDK 8 and OpenJFX installed. If it works for you, we're done with this issue.

**Comment by larsnaesbye** [comment]
@shirishag75 : can you confirm this works with OpenJDK 8 and OpenJFX? :)

**Comment by shirishag75** [comment]
@larsnaesbye and @mokun could we defer this till week-end. I have some free hours this week-end coming up where I hope to get some free space on one of other boxes, compile the whole thing under maven (hopefully with debugging fully turned on) and then see how it performs. I will send a mail (probably tomorrow) asking if there is complete documentation about how to compile under GNU/Linux and then try the same at my end.

**Comment by larsnaesbye** [comment]
Absolutely. Let us know if there are any challenges along the way.

**Comment by mokun** [comment]
@shirishag75, I've compiled r3899 into binary. see [v3.1.0-p6](https://sourceforge.net/projects/mars-sim/files/mars-sim/3.1.0/preview6/) Let me know.

To compile the source into jar/deb, we use maven. 

The easiest way in setting up the [developmental environment](https://github.com/mars-sim/mars-sim/wiki/Development-Environment) is with eclipse.

**Comment by mokun** [comment]
In short, we simply run the following in maven : 

> mvn clean package install

see https://github.com/mars-sim/mars-sim/wiki/Setting-Command-Line-Environment-via-SVN#3-building-and-running-mars-sim

I had a brief discussion with a user asking about how to set up maven properly in this [post](https://sourceforge.net/p/mars-sim/discussion/help/thread/0606e697/). 

There are plenty of posts out there with regarding to running maven in Eclipse or in command line.
For instance,  
http://stackoverflow.com/questions/15532534/maven-why-after-mvn-clean-i-need-to-execute-maven-update-project-before-m

**Comment by mokun** [comment]
See this eclipse snapshot in how to set up a maven configuration inside the "Run Configurations" panel below : 
![run maven in eclipse](https://cloud.githubusercontent.com/assets/1168584/24234033/6328a4d4-0f53-11e7-97a9-4084b63cb814.png)

**Comment by shirishag75** [comment]
@mokun - in Debian installing maven is a single line install - 

`$ sudo aptitude install maven` 

This is on a newer box, meaning better cpu, ram and works  - 

```
> mvn --version                                                                                                                   
Apache Maven 3.3.9
Maven home: /usr/share/maven
Java version: 1.8.0_121, vendor: Oracle Corporation
Java home: /usr/lib/jvm/java-8-openjdk-amd64/jre
Default locale: en_IN, platform encoding: UTF-8
OS name: "linux", version: "4.9.0-2-amd64", arch: "amd64", family: "unix"
```
I am in the process of updating trunk via - 

`$ svn up trunk `

will do the latter part once all the files have been added.

**Comment by mokun** [comment]
Yes I should have pointed out that when you run maven, it'll need to be at the top directory.

Also,  in mars-sim-main's pom.xml, we specify the following,

> 		<plugin>
>         		<groupId>org.apache.maven.plugins</groupId>
>         		<artifactId>maven-jar-plugin</artifactId>
>         		<version>2.6</version>
>         		<configuration>
>           			<archive>
>            				<manifest>
>             				<addClasspath>true</addClasspath>
>               				<mainClass>org.mars_sim.msp.MarsProjectStarter</mainClass>
>             			</manifest>
>           			</archive>
>         		</configuration>
>       		</plugin>

We actually use MarsProjectStarter beforehand to "prepare" for the JVM environment, such as setting up logging and enforcing specific memory setting, making it more suitable to run mars-sim on top of it. 

After that, we run MarsProjectFX to start mars-sim.

**Comment by shirishag75** [comment]
@all, trying out https://www.dropbox.com/s/23ou4cdpue8xpc5/3904.7z, will take a while to download and unzip and see if it works, sorry for taking so much time.

**Comment by mokun** [comment]
> the simulation doesn't tell/show anything 

@shirishag75, when a user chooses to use command line (instead of double-clicking the jar file) to start mars-sim, all the log statement will be available to him. 

It shows you started the sim at 6:53:27 PM and quit at 6:57:54 PM. 

What's the time ratio (aka speed) that your machine is using ? It's possible you are using the lowest at 128x. If you increase it to 2048x, it will simulate more frames per second. As time goes faster, you'll see the settlers attempt to do more things.

Should we print out log statements on everything that happens under the sim? 

For now,  we only display 2 levels, namely "INFO" and "SEVERE".  

> /home/shirish/games> java -jar 3904.jar                                                                                                         
> Command: java -Djava.util.logging.config.file=logging.properties -cp .:*:jars/* org.mars_sim.msp.javafx.MarsProjectFX -Xms256m -Xmx1024m -new
> OUTPUT>5 Apr, 2017 6:53:27 PM(INFO) main.javafx.MarsProjectFX Mars Simulation Project v3.1.0 - Build 3904 - 64-bit 1.8.0_121 - 4 CPU threads
> OUTPUT>5 Apr, 2017 6:53:27 PM(INFO) main.javafx.MarsProjectFX Starting a new sim in GUI mode in Linux
> OUTPUT>5 Apr, 2017 6:53:28 PM(INFO) ui.javafx.mainmenu.SpinningGlobe JavaFX 3D features supported
> OUTPUT>5 Apr, 2017 6:53:47 PM(INFO) core.time.MasterClock CPU Benchmark 1 : 51190 ms
> OUTPUT>5 Apr, 2017 6:53:47 PM(INFO) core.structure.Settlement Setting Alpha Base's Objective to Crop Farm
> OUTPUT>5 Apr, 2017 6:53:47 PM(INFO) core.structure.Settlement Setting Schiaparelli Point's Objective to Crop Farm
> OUTPUT>5 Apr, 2017 6:53:50 PM(SEVERE) core.time.UpTimer Too long since last time pulse, clearing update time
> OUTPUT>5 Apr, 2017 6:55:41 PM(INFO) core.person.PhysicalCondition Frank Crossman comes down with a flu
> OUTPUT>5 Apr, 2017 6:57:54 PM(INFO) ui.swing.MainDesktopPane Desktop update thread shutdown
> OUTPUT>5 Apr, 2017 6:57:54 PM(INFO) ui.javafx.MainScene Exiting the simulation. Bye!
>

**Comment by mokun** [comment]
> On the left-corner the white thing is the window that doesn't tell/show anything.

Can you tell us more about this white corner thing ? How long does it stay there ? You mention this for the first time.

It sounds like just an artifact, perhaps a JavaFX's Stage that Debian OS was trying to hold onto at the start of the sim instead of killing it. 

We can look for workaround to make sure JavaFX works better in Debian.

I didn't see that in Ubuntu.

**Comment by mokun** [comment]
> Downloaded and tried https://www.dropbox.com/s/9ilhvjpjlomie26/3927.jar.7z but no change :( . 

I made a lot changes since 3904.

> I did get sound through, without trying anything in the sound panel.

Sure. The exception/error is gone because I disabled the use of mars-sim Sound Panel in case of linux. This way, mars-sim won't try to seize control of sound volume and thus causing conflict with linux.

**Comment by shirishag75** [comment]
> What's the time ratio (aka speed) that your machine is using ? It's possible you are using the lowest at 128x. If you increase it to 2048x, it will simulate more frames per second. As time goes faster, you'll see the settlers attempt to do more things.

How do I do that ?

> Can you tell us more about this white corner thing ? How long does it stay there ? You mention this for the first time.

I mentioned it because these are the first times I managed to come to the tutorial part before this/today. 

It stays the whole time and it says/said Untitled window at the top, I thought it might be the simulation running but for some reason I'm not able to view it. Apart from the white untitled window I just see the tutorial window as shown above and that's that. 

> It sounds like just an artifact, perhaps a JavaFX's Stage that Debian OS was trying to hold onto at the start of the sim instead of killing it.

Maybe, but unless there is someway I can give you more data this would be mere speculation on both our parts. 

> We can look for workaround to make sure JavaFX works better in Debian.

sure, I'm all ears. 

> I didn't see that in Ubuntu.

Strange. hmm.... could it be because I'm running in 64-bit libraries  ?  The Ubuntu install that you tried was a 32-bit or 64-bit install ?

**Comment by mokun** [comment]
> How do I do that ?

Click on the Speed Panel on top right hand corner. Drag the slider bar all the way to the right. Time ratio will increase to max.

**Comment by mokun** [comment]
> mere speculation on both our parts.

Let's just ignore that window for a second. It's not that important. you can still run mars-sim. 

Each JavaFX Stage is considered a separate application and thus having a separate icon. It's weird Java VM behaves this way but it's by default.

I know for fact that in Ubuntu, there's also a bug in Java API that prevents from displaying an stylish icon on the Ubuntu bar on the left of the screen...

> Apart from the white untitled window I just see the tutorial window as shown above and that's that.

You should be able to see the 3 choices (Main, Map, Help) on the Tab bar.

If the view of them were blocked by that white box, then drag the mars-sim window to the right (by hovering the mouse over the title bar and drag the mouse).

> Ubuntu install that you tried was a 32-bit or 64-bit install 

I have 64-bit ubuntu 17.04. 

Also, I've using OpenJDK 8 with OpenJFX

I assume you are using with OpenJDK/OpenJFX, is it possible for you to switch to Oracle JDK 8 to see if the white box goes away ?

**Comment by Musickiller** [comment]
worked -well- more or less no my machine... just so that you knew that it's not a worldwide problem..

**Comment by shirishag75** [comment]
aha, was able to get the tab bar as well as click on the map and got the settlement map, But found it very unintuitive. Also there doesn't seem to be a way to know what is happening and if something new is being constructed or not. I did see notifications but they were too low and went on too fast to be of any consequence. Maybe going forward, we might have also have a notifications bar, who knows ;)

**Comment by mokun** [comment]
> there was however a white box whenever a pause window appeared.

@Musickiller , so the white box doesn't show up at the start of the sim. Rather, it's when the user hit ESC to pause the sim, right ?

btw, I'll work on adding back a pause "||" button.

**Comment by mokun** [comment]
> But found it very unintuitive. 

@shirishag75 , sure. I agree we should figure out how and what to do to make it as intuitive as possible as we can. 

> Also there doesn't seem to be a way to know what is happening and if something new is being constructed or not. 

Currently, mars-sim tracks/records only two types of events, namely, equipment malfunction and medical emergency. 

They are listed under the data table in the "Events" tab in the Monitor Tool.

We could capture new building and new construction events and use a filter mechanism to sort and display more types of events in that data table in future.

**Comment by mokun** [comment]
> I did see notifications but they were too low and went on too fast to be of any consequence. 

Note that the notification boxes popping up on the bottom-left are equipment malfunctions while the boxes popping up on the bottom-right are medical related injuries. 

I could lengthen the timing to let them stay longer but all the malfunction and medical events are captured in the event data table anyway.

One caveat is that when someone saved and end the sim, all the previous events are gone even when you load the sim back up. They cannot be saved. The past developers somehow considered it unnecessary to save the past events. 

One possible reason is that it can waste a lot of stack space in serializing the HistoricalEvent and HistoricalManager class. But that was before we implemented the xz compression scheme that effectively reduces the size of the saved sim by 5 to 10 times. 

Having said that, we could redesign those class so that events can be saved into separate json files, etc. in future.

**Comment by mokun** [comment]
> Maybe going forward, we might have also have a notifications bar

I've got a bottom notification bar at some point in the past. We'll need to quantify exactly what type of events we want it to report. 

To me, we could use it to report resources shortage and so forth.

May be it's just me, but my experience with the last few Civilization games is that once I get a hang of the game mechanics and what to do to next, I'm very busy with what I need to do and pretty much turn off all the notifications. They are too distracting and are no longer helpful. If I need something, I'll go straight to where the data is presented.

**Comment by shirishag75** [comment]
> Currently, mars-sim tracks/records only two CATEGORIES of events, namely, (1) equipment malfunction and (2) medical emergency.

> They are listed under the data table in the "Events" tab in the Monitor Tool.

This unearthed another bug which I have reported separately, https://github.com/mars-sim/mars-sim/issues/46 . We shouldn't be dumping all the bugs/issues on this massive bug-report only.

**Comment by shirishag75** [comment]
> Note that the notification boxes popping up on the bottom-left are equipment malfunctions while the boxes popping up on the bottom-right are medical related injuries.
> 
> I could lengthen the timing to let them stay longer but all the malfunction and medical events are captured in the event data table anyway.
> 
> One caveat is that when someone saved and end the sim, all the previous events are gone even when you load the sim back up. They cannot be saved. The past developers somehow considered it unnecessary to save the past events.
> 
> One possible reason is that it can waste a lot of stack space in serializing the HistoricalEvent and HistoricalManager class. But that was before we implemented the xz compression scheme that effectively reduces the size of the saved sim by 5 to 10 times.
> 
> Having said that, we could redesign those class so that events can be saved into separate json files, etc. in future.

That would be much desirable and also to know/understand what happened previously.

**Comment by shirishag75** [comment]
> I've got a bottom notification bar at some point in the past. We'll need to quantify exactly what type of events we want it to report.
> 
> To me, we could use it to report resources shortage and so forth.
> 
> May be it's just me, but my experience with the last few Civilization games is that once I get a hang of the game mechanics and what to do to next, I'm very busy with what I need to do and pretty much turn off all the notifications. They are too distracting and are no longer helpful. If I need something, I'll go straight to where the data is presented.

There is a difference as rightly pointed out by you when the user knows or has enough knowledge of the game mechanics that he turns off what is unnecessary and a newbie who doesn't know what s(he) has come onto. For her(im) providing as much data and understanding is good at the beginning, once s(he) is familiar with the interface and knows what to expect s(he) would automatically do the needful. 

Ideally, we should have short tutorials introducing parts of the game mechanic to the player so s(he) has an idea of what s(he) is supposed to do. Right now, it is all a mystery to me.

**Comment by mokun** [comment]
> For her(im) providing as much data and understanding is good at the beginning, once s(he) is familiar with the interface and knows what to expect s(he) would automatically do the needful.

Agree. 

> Ideally, we should have short tutorials introducing parts of the game mechanic to the player so s(he) has an idea of what s(he) is supposed to do. Right now, it is all a mystery to me.

Yes, a good, interactive tutorial will surely help a lot in introducing mars-sim to newbies.

Our line of thinking is to rely on the Help Browser to provide the tutorial. It's possible that we do it without using the Help Browser, which is currently accessible on a different tab. Prior to v3.1.0, all panels exist in one big desktop area, including the help panel. Now, We use 3 separate tabs instead.

At some point, we'll integrate a new help system in mars-sim.

**Comment by mokun** [comment]
> there was however a white box whenever a pause window appeared. have not tried when you've removed it.

@Musickiller , I've since removed the white box that you pointed out.

see [r3962](https://github.com/mars-sim/mars-sim/commit/40a56f2acded217f02ddc026600271ac24bced45)

**Comment by mokun** [comment]
There are multiple issues on this thread and they have been mostly worked on. Closing this for now.
